src/combinations.rs

```diff
@@ -103,41 +153,42 @@ where
 {
     type Item = Vec<I::Item>;
     fn next(&mut self) -> Option<Self::Item> {
-        if self.first {
-            self.pool.prefill(self.k());
-            if self.k() > self.n() {
-                return None;
-            }
-            self.first = false;
-        } else if self.indices.is_empty() {
-            return None;
+        let done = if self.first {
+            self.init()
         } else {
-            // Scan from the end, looking for an index to increment
-            let mut i: usize = self.indices.len() - 1;
+            self.increment_indices()
+        };
 
-            // Check if we need to consume more from the iterator
-            if self.indices[i] == self.pool.len() - 1 {
-                self.pool.get_next(); // may change pool size
-            }
+        if done {
+            return None;
+        }
 
-            while self.indices[i] == i + self.pool.len() - self.indices.len() {
-                if i > 0 {
-                    i -= 1;
-                } else {
-                    // Reached the last combination
-                    return None;
-                }
-            }
+        Some(self.indices.iter().map(|i| self.pool[*i].clone()).collect())
+    }
 
-            // Increment index, and reset the ones to its right
-            self.indices[i] += 1;
-            for j in i + 1..self.indices.len() {
-                self.indices[j] = self.indices[j - 1] + 1;
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        if n == 0 {
+            return self.next();
+        }
+
+        let mut done = if self.first {
+            self.init()
+        } else {
+            self.increment_indices()
+        };
+
+        if done {
+            return None;
+        }
+
+        for _ in 0..(n - 1) {
+            done = self.increment_indices();
+            if done {
+                return None;
             }
         }
 
-        // Create result vector based on the indices
-        Some(self.indices.iter().map(|i| self.pool[*i].clone()).collect())
+        self.next()
```

Comment:
Should we not iterate `0..n` and return `Some(...collect())` rather `self.next()` which has `init` code we don't need at this point, right?

I assume it was not to duplicate the `indices.iter...collect` part.
We could (in another PR) have one helper function to do this `indices.iter...collect` which would be basically used in all iterators returning vector items.

Reply:
Yep that was my thinking and yeah I make that change to skip the extra boolean comparison if you like?

Reply:
I don't see the "extra boolean comparison".
If I inlined this last `next`, I'd get a `let done = if first { init /*won't get here, might as well delete it*/} else { increment }; if done {return None} Some(collect)` so it basically completes the `0..n-1` to `0..n` and then return some.

Reply:
Ah yeah I mean the way I currently have it has an extra comparison. Happy to change it if you like?

Reply:
Yes I'd like this change.